BaikalDataGenerator/Source/data_generator_impl.cpp

```diff
@@ -0,0 +1,326 @@
+    int channels_num;
+};
+
+namespace {
+
+// if you need to add new output for saving to disk
+// for iterations number counted in spp.xml file
+// just put its description in this collection
+const std::vector<OutputInfo> kMultipleIteratedOutputs
+{
+    {Baikal::Renderer::OutputType::kColor, "color", 3},
+    {Baikal::Renderer::OutputType::kAlbedo, "albedo", 3},
+    {Baikal::Renderer::OutputType::kGloss, "gloss", 1}
+};
+
+// if you need to add new output for saving to disk
+// only for the one time
+// just put its description in this collection
+const std::vector<OutputInfo> kSingleIteratedOutputs
+{
+    {Baikal::Renderer::OutputType::kViewShadingNormal, "view_shading_normal", 3},
+    {Baikal::Renderer::OutputType::kDepth, "view_shading_depth", 1}
+};
+
+} // namespace
+
+
+DataGeneratorImpl::DataGeneratorImpl(SceneObject* scene,
+                                     size_t output_width,
+                                     size_t output_height,
+                                     std::uint32_t num_bounces,
+                                     unsigned device_idx)
+    : m_scene(scene),
+      m_width(static_cast<std::uint32_t>(output_width)),
+      m_height(static_cast<std::uint32_t>(output_height)),
+      m_num_bounces(num_bounces),
+      m_device_idx(device_idx)
+{
+    assert(m_width);
+    assert(m_height);
+    assert(num_bounces);
+
+    auto devices = GetDevices();
+    if (devices.empty())
+    {
+        THROW_EX("Cannot find any device");
+    }
+    if (device_idx >= devices.size())
+    {
+        THROW_EX("Cannot find device with index " << device_idx);
+    }
+
+    m_context = std::make_unique<CLWContext>(CLWContext::Create(devices[device_idx]));
+
+    m_factory = std::make_unique<Baikal::ClwRenderFactory>(*m_context, "cache");
+
+    auto render = m_factory->CreateRenderer(
+        Baikal::ClwRenderFactory::RendererType::kUnidirectionalPathTracer);
+    m_renderer.reset(dynamic_cast<Baikal::MonteCarloRenderer*>(render.release()));
+
+    m_controller = m_factory->CreateSceneController();
+
+    for (auto& output_info : kMultipleIteratedOutputs)
+    {
+        m_outputs.push_back(m_factory->CreateOutput(m_width, m_height));
+        m_renderer->SetOutput(output_info.type, m_outputs.back().get());
+    }
+    for (auto& output_info : kSingleIteratedOutputs)
+    {
+        m_outputs.push_back(m_factory->CreateOutput(m_width, m_height));
+        m_renderer->SetOutput(output_info.type, m_outputs.back().get());
+    }
+
+    m_renderer->SetMaxBounces(m_num_bounces);
+}
+
+void DataGeneratorImpl::AttachLight(LightObject* light)
+{
+    m_scene->AttachLight(light);
+}
+
+void DataGeneratorImpl::SaveMetadata(const std::filesystem::path& output_dir,
+                                     const std::string& scene_name,
+                                     unsigned cameras_start_idx,
+                                     unsigned cameras_end_idx,
+                                     int cameras_index_offset,
```

Comment:
cameras_start_idx, cameras_end_idx and cameras_index_oofset are not used here. Please, remove them

Reply:
Done